[ci skip] 1532213: Add inline metrics docs (#2544)
mozilla-mobile/android-components: components/service/glean/docs/metrics/adding-new-metrics.md

When adding a new metric, the workflow is:
* Decide on which metric type you want to use.
* Add a new entry to `metrics.yaml`.
* Add code to your project to record into the metric by calling glean.

**Important**: as stated [here](../../README.md#before-using-the-library), any new data collection requires
documentation and data-review. This is also required for any new metric automatically collected
by glean.

> # The `metrics.yaml` file

Reviewer: I feel this whole section is too long at this place and this doc is not snappy anymore.
I may be off here, so feel free to counter.

Could we...
- Move this further down?
   - after the "choosing a metric type"? My assumption being that looking for a metric type to match their problem is what people come here first with, then they want details on how that works in detail.
   - Or instead have one short section on "getting kickstarted with the metrics.yaml file" through one hands-on example... and put the other details here further down or in a separate file?
- Add a top-level TOC around here or further up?

Author: What I was really trying to address here is that these docs didn't cover any of the required metric parameters (they just leave them out with `...`) which makes it hard to just dive in.  Maybe the thing to do is to have a single complete example near the top here and move the details about each parameter to the bottom?

Author: I've fleshed out the main example, and moved these details to the bottom.

The `metrics.yaml` file defines the metrics your application or library will
send. They are organized into categories. Category names may optionally contain
`.`'s to indicate subcategories. 

Therefore, the overall organization is:

```YAML
# Required to indicate this is a `metrics.yaml` file
$schema: moz://mozilla.org/schemas/glean/metrics/1-0-0

category1:
  metric1:
    ...
  metric2:
    ...
    
category2.subcategory:
  metric3:
    ...

```

## Metric naming

Category and metric names in the `metrics.yaml` are in `snake_case`, but given
the Kotlin coding standards defined by
[ktlint](https://github.com/pinterest/ktlint), these identifiers must be
`camelCase` in Kotlin. Therefore, when the Kotlin wrappers are generated the
`snake_case` identifiers are automatically converted to `camelCase`. Be aware of
that when using the examples below.

# Common metric parameters

All metric types must include the following required parameters:

> - `description**: **Required.** A textual description of the metric for humans.

Reviewer 2: ```suggestion
- `description`: **Required.** A textual description of the metric for humans.
```

  - `<build date>`: An ISO date `yyyy-mm-dd` in UTC on which the metric expires.
    For example, `2019-03-13`. This date is checked at build time. Except in
    special cases, this form should be used so that the metric automatically
    "sunsets" after a period of time.
  - `never`: This metric never expires.
  - `expired`: This metric is manually expired.
  
All metric types also support the following optional parameters:

- `lifetime`: Defines the lifetime of the metric. Different lifetimes affect
  when the metrics value is reset.
  - `ping` (default): The metric is reset each time it is sent in a ping.
  - `user`: The metric is part of the user's profile.
  - `application`: The metric is related to an application run, and is reset
    when the application restarts.
    
- `send_in_pings`: Defines which pings the metric should be sent on. If not
  specified, the metric is sent on the "default ping", which is the `events`
  ping for events and the `metrics` ping for everything else. Most metrics don't
  need to specify this unless they are sent on custom pings.
  
- `disabled`: (default: `false`) Data collection for this metric is disabled.

> - `version`: (default: 0) The version of the metric. A monotonically increasing

Reviewer 2: ```suggestion
- `version`: (default: `0`) The version of the metric. A monotonically increasing
```

  integer value. This should be bumped if the metric changes in a
  backward-incompatible way.
  
For clarity, these common parameters are ommitted from the examples about
specific metric types below.

# Choosing a metric type

There are different metrics to choose from, depending on what you want to achieve:

> * [Events](#Events): Records events e.g. individual occurences of user actions, say every time a view was open and from where.

Reviewer 2: ```suggestion
* [Events](#Events): Records events e.g. individual occurences of user actions, for example every time a view was opened and from where.
```

* [Booleans](#Booleans): Records a single truth value, for example "is a11y enabled?"
* [String](#String): Records a single Unicode string value, for example the name of the OS.
* [String lists](#String-lists): Records a list of Unicode string values, for example the list of enabled search engines.
* [Counters](#Counters): Used to count how often something happens, for example, how often a certain button was pressed.

> * [Timespans](#Timespans): Used to measure how much time is spent in a single task.

Reviewer 2: ```suggestion
* [Timespans](#Timespans): Used to measure time, for example how much time was spent in a single task.
```

* [Timing Distributions](#Timing-Distributions): Used to record the distribution of multiple time measurements.
* [Datetimes](#Datetimes): Used to record an absolute date and time, such as the time the user first ran the application.
* [UUIDs](#UUIDs): Used to record universally unique identifiers (UUIDs), such as a client ID.
* [Labeled metrics](#Labeled-metrics): Used to record multiple metrics of the same type under different string labels, say every time you want to get a count of different error types in one metric.

## Events

Events allow recording of e.g. individual occurences of user actions, say every time a view was open and from where.

> Each time you record an event, it records a timestamp, the event's name and a set of custom values.

Reviewer 2: ```suggestion
Each time you record an event it records a timestamp, the event's name, and a set of custom values.
```
Oxford comma :)

When adding a new metric, the workflow is:
* Decide on which metric type you want to use.
* Add a new entry to `metrics.yaml`.

> * Add code to your project to record into the metric by calling glean.

Reviewer: Can we linkify this list for easy access? Or add a toplevel TOC?

Author: Where would these links point at?  Depending on the event type, we have a bunch of targets for these, but not a single one...

Reviewer: Yeah, maybe it's hard to fit this into this list.

I think currently nothing allows me to jump directly to `# Adding new metrics`, `# The `metrics.yaml` file`, `# Choosing a metric type`.
Adding these 3 as a TOC at the start of this doc?

But let's not block on this, happy to talk next week for a follow-up or so?

Reviewer: Also: `# Common metric parameters`.

**Important**: as stated [here](../../README.md#before-using-the-library), any new data collection requires
documentation and data-review. This is also required for any new metric automatically collected
by glean.

# The `metrics.yaml` file

The `metrics.yaml` file defines the metrics your application or library will
send. They are organized into categories. Category names may optionally contain

> `.`'s to indicate subcategories. 

Reviewer: Nit: drop the subcategories here.

Author: I find that feature *really* surprising, and I don't think there's another way to discover it if we don't document it.

Reviewer: But if you show it in the example, is that not sufficient?
But either way i'm fine with this.

Therefore, the overall organization is:

```YAML
# Required to indicate this is a `metrics.yaml` file
$schema: moz://mozilla.org/schemas/glean/metrics/1-0-0

category1:

>   metric1:

Reviewer: Can you stick with a real-ish, relatable example here?
Maybe an event, because we expect most initial instrumentation to be events.

    type: string
    description: |
      An example metric
    notification_emails:
      - CHANGE-ME@test-only.com
    bugs:
      - 123456789
    data_reviews:
      - http://test-only.com/path/to/data-review
    expires:

>       - 2100-01-01

Reviewer: Definitely use a sooner date or something that doesn’t build here to avoid copy-pasting of this far-off date.

Author: This is a hard one.  If we make it a sooner date, then copying-and-pasting will result in an already disabled event.  Probably have to use a comment here to avoid confusion.

Reviewer: Yeah, maybe... an expired date and a comment to change it?

    
  metric2:
    ...
    
category2.subcategory:
  metric3:
    ...

```

The details of the metric parameters are described below in [Common metric
parameters](#Common-metric-parameters).

## Metric naming

Category and metric names in the `metrics.yaml` are in `snake_case`, but given
the Kotlin coding standards defined by
[ktlint](https://github.com/pinterest/ktlint), these identifiers must be

> `camelCase` in Kotlin. Therefore, when the Kotlin wrappers are generated the

Reviewer: Instead of “Therefore...”, can you put an example?

When adding a new metric, the workflow is:
* Decide on which metric type you want to use.
* Add a new entry to `metrics.yaml`.
* Add code to your project to record into the metric by calling glean.

**Important**: as stated [here](../../README.md#before-using-the-library), any new data collection requires
documentation and data-review. This is also required for any new metric automatically collected
by glean.

# The `metrics.yaml` file

The `metrics.yaml` file defines the metrics your application or library will
send. They are organized into categories. Category names may optionally contain
`.`'s to indicate subcategories. 

Therefore, the overall organization is:

```YAML
# Required to indicate this is a `metrics.yaml` file
$schema: moz://mozilla.org/schemas/glean/metrics/1-0-0

> toolbar:

Reviewer: And could have a comment here on the categories?
Hands-on explanations... but feel free to disregard if that's not your approach.

  click:
    type: event
    description: |
      Event to record toolbar clicks.
    notification_emails:
      - CHANGE-ME@test-only.com
    bugs:
      - 123456789
    data_reviews:
      - http://test-only.com/path/to/data-review
    expires:
      - 2019-06-01  # <-- Update to a date in the future
    
  metric2:
    ...
    

> category2.subcategory:

Reviewer: Could just have a comment here on the subcategories?

  metric3:
    ...

```

The details of the metric parameters are described below in [Common metric
parameters](#Common-metric-parameters).

## Metric naming

Category and metric names in the `metrics.yaml` are in `snake_case`, but given

> the Kotlin coding standards defined by

Reviewer: I think we haven't mentioned yet that the Kotlin APIs are generated from this?
Fine for a follow-up :)

Author: I added a single sentence to the `metrics.yaml` file section that hopefully covers this.

[ktlint](https://github.com/pinterest/ktlint), these identifiers must be
`camelCase` in Kotlin. For example, the metric defined in the `metrics.yaml` as:

```YAML
views:
  login_opened:
    ...
```

is accessible in Kotlin as:

```Kotlin
import org.mozilla.yourApplication.GleanMetrics.Views
Views.loginOpened...
```

# Choosing a metric type

There are different metrics to choose from, depending on what you want to achieve:

* [Events](#Events): Records events e.g. individual occurences of user actions, say every time a view was open and from where.
* [Booleans](#Booleans): Records a single truth value, for example "is a11y enabled?"
* [Strings](#Strings): Records a single Unicode string value, for example the name of the OS.

> * [String lists](#String-lists): Records a list of Unicode string values, for example the list of enabled search engines.

Reviewer 2: nit: 
```suggestion
* [String Lists](#String-lists): Records a list of Unicode string values, for example the list of enabled search engines.
```

* [Counters](#Counters): Used to count how often something happens, for example, how often a certain button was pressed.
* [Timespans](#Timespans): Used to measure how much time is spent in a single task.
* [Timing Distributions](#Timing-Distributions): Used to record the distribution of multiple time measurements.
* [Datetimes](#Datetimes): Used to record an absolute date and time, such as the time the user first ran the application.
* [UUIDs](#UUIDs): Used to record universally unique identifiers (UUIDs), such as a client ID.

> * [Labeled metrics](#Labeled-metrics): Used to record multiple metrics of the same type under different string labels, say every time you want to get a count of different error types in one metric.

Reviewer 2: nit: 
```suggestion
* [Labeled Metrics](#Labeled-metrics): Used to record multiple metrics of the same type under different string labels, say every time you want to get a count of different error types in one metric.
```

import org.mozilla.yourApplication.GleanMetrics.Rendering

Rendering.imageDecodeTime.start() // start the timer
try {
   // ... decode an image ...
} catch (e: Exception) {
   Rendering.imageDecodeTime.cancel() // cancel the timer -- nothing is recorded
}
Rendering.imageDecodeTime.stopAndSum() // stop the timer
```

The time reported in the telemetry ping will be the sum of all of these
timespans recorded during the lifetime of the ping.

There are test APIs available too:

```Kotlin
import org.mozilla.yourApplication.GleanMetrics.Rendering
Glean.enableTestingMode()

// Was anything recorded?
assertTrue(Rendering.imageDecodeTime.testHasValue())
// Does the timer have the expected value
assertTrue(Rendering.imageDecodeTime.testGetValue() > 0)
```

## Timing Distributions

Timing distributions are used to accumulate and store time measurement, for analyzing distributions of the timing data.

### Metric-specific parameters

Timing distributions have a required `time_unit` parameter to specify the resolution and range of the values that are recorded. The allowed values for `time_unit` are:

   `nanosecond`, `microsecond`, `millisecond`, `second`, `minute`, `hour`, and
   `day`

> Which range 

Reviewer 2: nit: remove the line break here

### Metric-specific properties

Datetimes have a required `time_unit` parameter to specify the smallest unit
of resolution that the timespan will record. The allowed values for `time_unit` are:

   `nanosecond`, `microsecond`, `millisecond`, `second`, `minute`, `hour`, and
   `day`

Carefully consider the required resolution for recording your metric, and choose
the coarsest resolution possible.

### Usage

You first need to add an entry for it to the `metrics.yaml` file:

```YAML
install:
  first_run:
    type: datetime 
    time_unit: day 
    description: >
      Records the date when the application was first run
    ...
```

Now that the datetime is defined in `metrics.yaml` you can use the metric to
record datetimes in the application's code:

```Kotlin
import org.mozilla.yourApplication.GleanMetrics.Install

Install.firstRun.set() // Records "now"
Install.firstRun.set(Calendar(2019, 3, 25)) // Records a custom datetime
```

There are test APIs available too.

```Kotlin
import org.mozilla.yourApplication.GleanMetrics.Install
Glean.enableTestingMode()

// Was anything recorded?
assertTrue(Install.firstRun.testHasValue())
// Was it the expected value? 

> assertEquals("2019-03-25-05:00", Install.firstRun.testGetValueAsString())

Reviewer 2: nit: Your defined metric above has a time unit of `day` and this is comparing as if it were `hour`

Author: The `-05:00` bit is actually the timezone offset.  If there were a time, it would look like `2019-03-25T09:15:00-05:00`.  I agree this is confusing -- maybe I'll add a comment to this effect.

Reviewer 2: Doh!  Makes sense now that I look at it with that in mind ;)

// Usually you don't know the exact timing values, but how many should have been recorded.
assertEquals(2L, snapshot.count())
```

## Datetimes

Datetimes are used to record an absolute date and time, for example the date and time that the application was first run.

You first need to add an entry for it to the `metrics.yaml` file:

```YAML
install:
  first_run:
    type: datetime 
    time_unit: day 
    description: >
      Records the date when the application was first run
    ...
```

Datetimes have a required `time_unit` parameter to specify the smallest unit
of resolution that the timespan will record. The allowed values for `time_unit` are:

   `nanosecond`, `microsecond`, `millisecond`, `second`, `minute`, `hour`, and
   `day`

Carefully consider the required resolution for recording your metric, and choose
the coarsest resolution possible.

The device's offset from UTC is recorded and sent with the datetime value in the ping.

Now that the datetime is defined in `metrics.yaml` you can use the metric to
record datetimes in the application's code:

```Kotlin
import org.mozilla.yourApplication.GleanMetrics.Install

> // ...

Reviewer 3: nit: let's remove `//...` if we have nothing to add :)

assertTrue(User.clientId.testHasValue())
// Was it the expected value? 
assertEquals(uuid, User.clientId.testGetValue())
```

## Labeled metrics

Some metrics can be used as *labeled* variants. This means that for a single metric entry you define in `metrics.yaml`, you can record into multiple metrics under the same name, each identified by a different string label.

This is useful when you need to break down metrics by a label known at build time or run time. For example:
- When you want to count a different set of subviews that users interact with, you could use `viewCount["view1"].add()` and `viewCount["view2"].add()`.
- When you want to count errors that might occur for a feature, you could use `errorCount[errorName].add()`.

**Note**: Be careful with using arbitrary strings as labels and make sure they can't accidentally contain identifying data (like directory paths or user input).

Say you're adding a new counter for errors that can occur when loading a resource from a REST API. First you need to add an entry for the counter to the `metrics.yaml` file:

```YAML
updater:
  load_error:
    type: counter
    labeled: true # This makes it a labeled counter.

>     labels: # This is optional, if provided it limits the set of labels you can use.

Reviewer 3: I would make this stand out and encourage users to use `labels`, unless labels are not known beforehand.
